### ISSUE TITLE
Introduce tooltips for better contextual help

### ISSUE THREAD

**Issue by michaelbeil** [AffiliateWP/AffiliateWP#1412]
To help reduce support, we should introduce inline, contextual tooltips throughout AffiliateWP per [RCP](https://github.com/restrictcontentpro/restrict-content-pro/issues/518) and [EDD](https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4641).   

We can use jQuery UI Tooltip which is included in WordPress core and simply port over what has been done in [EDD](https://github.com/easydigitaldownloads/easy-digital-downloads/search?utf8=%E2%9C%93&q=tooltip) and [RCP](https://github.com/restrictcontentpro/restrict-content-pro/search?utf8=%E2%9C%93&q=tooltip).

**Comment by DrewAPicture** [comment]
This is going to be a huge undertaking and I don't think it's going to happen for 2.0.

**Comment by ramiabraham** [comment]
@AffiliateWP/core Added a PR: https://github.com/AffiliateWP/AffiliateWP/pull/2385.

Presently, markup is rendered twice for some reason. Otherwise this is ready for review.

**Comment by ramiabraham** [comment]
To test, add the following to any core AffiliateWP setting array:

```php
'tooltip_title'   => 'Here is the tooltip title',
'tooltip_desc' => 'Here is the tooltip description',
```

---

- Each core setting type now echoes a filter instead of just the `$html`. The filter passes `$html` and `$args`:

```php
// (some core callback)
echo apply_filters( 'affwp_after_setting_output', '<hr/>', $args );
```

- Question: Should there be a standalone function to render a tooltip inline? Such as `affwp_tooltip( $title, $desc )`?

**Comment by DrewAPicture** [comment]
How accessible are these tooltips?

**Comment by pippinsplugins** [comment]
@DrewAPicture if I recall, we had them reviewed for accessibility when we introduced them into EDD and they passed. This uses the same method.

**Comment by ramiabraham** [comment]
@DrewAPicture @pippinsplugins (referenced in chat, [1](https://eddwp.slack.com/archives/G0EL3RQH2/p1505167531000401), [2](https://eddwp.slack.com/archives/G0EL3RQH2/p1505167736000151), [3](https://eddwp.slack.com/archives/G0EL3RQH2/p1505169247000174))

**Comment by pippinsplugins** [comment]
@ramiabraham yep I'm happy with this. Could you please go ahead and make a pass on adding the tooltips themselves?

**Comment by ramiabraham** [comment]
@AffiliateWP/core Sweet! Next question: What things do we want tooltips for?

**Comment by pippinsplugins** [comment]
@ramiabraham I'd like tooltips for nearly all of the settings and any interface element that is not abundantly clear.

**Comment by ramiabraham** [comment]
An update:

- Resolved the duplicate output issue in https://github.com/AffiliateWP/AffiliateWP/commit/60d40b8823f15fec9d90826a34c2d5aa9cfabdc9
- Added the first of the tooltips in https://github.com/AffiliateWP/AffiliateWP/commit/9ac90a71db013735ab6ce82b04f2adbae08b5812 - wording look ok?
- I'm seeing `sprintf` breaking markup at the close of `<a href="%s">`, working on that as well.
- Continuing with _tooltipping all the things_

**Comment by pippinsplugins** [comment]
@ramiabraham left a few suggestions for you.

**Comment by pippinsplugins** [comment]
Looking good, @ramiabraham! Let's go ahead and get tooltips added to the other settings too.

**Comment by ramiabraham** [comment]
@AffiliateWP/core Added _much_ more in https://github.com/AffiliateWP/AffiliateWP/commit/036b6a68d07fe04da55a56797f5203a07838c60a - any we're missing?

**Comment by pippinsplugins** [comment]
@ramiabraham I see a few things that need adjusted still.

1. The tool tips need to be inline, not on new lines:
<img width="1031" alt="screen shot 2017-10-05 at 1 15 53 pm" src="https://user-images.githubusercontent.com/1034109/31244070-64c8bf24-a9cf-11e7-800a-d5099d17ea22.png">

2. There is an erroneous space after `[affiliate_area]`:
<img width="258" alt="screen shot 2017-10-05 at 1 16 46 pm" src="https://user-images.githubusercontent.com/1034109/31244593-81962db2-a9cf-11e7-833c-a6b092faa36a.png">

3. The `Referral Variable` tool tip is not very helpful. It should explain what it is and where to learn more, not just provide a link to learn more.

4. `Pretty Affiliate URLs` needs to have an in-depth tooltip

5. `Referral Rate` needs a tool tip that provides examples.

6. `Exclude Shipping` is using the `Credit Last Referrer`  tool tip.

7. `Exclude Taxes` needs a tooltip.

8. `Currency` needs a tooltip.

9. `Registration form fields` needs a tool tip.

10. The new email settings for disabling individual emails need tooltips.

11. `Reject Unpaid Referrals on Refund` needs a tooltip.

12. `Referral URL Blacklist` needs a tooltip.

13. `Opt into Beta Versions` needs a tooltip

14. `Remove Data on Uninstall` needs a tooltip.

**Comment by DrewAPicture** [comment]
So, I guess it begs the question of what the difference is between displaying contextual help all the time below a setting and what goes into a tooltip? And at what point have we made our settings so cryptic that almost all of them need extra text to explain what they are?

**Comment by DrewAPicture** [comment]
So, I guess it begs the question of what the difference is between displaying contextual help all the time below a setting and what goes into a tooltip?

And at what point have we made our settings so cryptic that almost all of them need extra text to explain what they are? Meaning that, at a certain point, perhaps we're working around the actual problem that our settings can't effectively explain themselves, either because the format (tabs of lists of settings) is flawed, or something else?

Did we ever actually define the problem we're trying to solve here? :-)

**Comment by pippinsplugins** [comment]
I think the same question can be asked about documentation? If the settings are hard to explain, why does documentation exist?

Unfortunately I don't think it matters how clearly we label settings, there will _always_ be someone who has difficulty in understanding them. The tooltips essentially act as a replacement/supplemental to the more verbose documentation that lives on our site and are solely for the individuals that need a more in-depth explanation of the individual options.

By "hiding" the in-depth information away in tool tips we both avoid content-overload on the page while also assisting those users do that want more information.

**Comment by DrewAPicture** [comment]
Actually defining the problem is sometimes hard. We know symptoms of the problem, for instance:

1. The settings don't adequately explain themselves
2. The setting names are unclear
3. The setting name and description don't match
4. The setting name is succinct (and therefore confusing because of it)
5. There are too many settings
6. There is little to no explanation of how settings interact with each other

But those aren't actually what's wrong, just indicators of a larger problem.

Here's my attempt at defining the problem: When multiple individual settings are combined, they don't adequately set expectations

So how does that fit into the tooltips discussion, exactly? I think in some of the most effective UIs I've ever seen, tooltips primarily help to selectively set expectations at a more granular level.

So, imho, the first thing we should probably do is breakdown each of our core settings and do the following:

* Write non-succinct descriptions for what they actually do
* Write down what other settings they directly influence
* Define the expectations once each setting is configured.

It might seem like a lot of work, but the value in doing this is that we'll pretty quickly see where the setting names and descriptions are succinct for the sake of succinctness, at the peril of understanding. I also think think we might discover that reorganizing some things more logically might be of particular benefit to more adequately setting expectations for how those settings will affect the overall experience.

Once we can get a handle on ensuring our base microcopy and organization are on point, then we should look at where tooltips can help make things clear.

"But I really feel like having tooltips now would be beneficial regardless". Maybe. The problem with approaching this as "we have a problem we've solved elsewhere before" is that it's not true :-) You've solved the problem something completely unrelated before, which shouldn't really have bearing here.

**Comment by pippinsplugins** [comment]
That's all very good.

In that vein then, I think we should start by defining which settings are problematic and apply @DrewAPicture's steps there and see where that leads us.

**Comment by ramiabraham** [comment]
Will start by working on the additional tooltips noted, and fleshing out some existing ones - I love these points. Going to use them as guidelines for each as @pippinsplugins noted:

> Write non-succinct descriptions for what they actually do
> Write down what other settings they directly influence
> Define the expectations once each setting is configured.

This will greatly increase the usability of our 🏍

**Comment by michaelbeil** [comment]
Is there an update for this that I should test? I don't see anything recently.

**Comment by michaelbeil** [comment]
Alright. Is there anything that I should test still?

**Comment by pippinsplugins** [comment]
No.

**Comment by michaelbeil** [comment]
Okay, thanks for letting me know.
